File: Coding Ninjas/Easy/Next smaller Palindrome.py
def nextSmallerPalindrome( s ):
    
    if s == "11":
        return "9"
    elif int( s ) < 11:
        return str( int( s ) - 1 )
        
    n = len( s ) // 2
    m = len( s ) % 2
    
    if s[ n - 1 : : -1 ] < s[ n + m : ]:
        return s[ : n ] + s[ n - ( 1 - m ) : : -1 ]
    else:
        t = str( int( s[ : n + m ] ) - 1 )

        if len( t ) < n + m:
            return "9" * ( len( s ) - 1 )
            
        return t + t[ n - m : : -1 ]

File: Coding Ninjas/Easy/Next smaller Palindrome.java
public class Solution
{

	public static String nextSmallerPalindrome
	( String s )
	{
		if ( s.equals( "1" ) )
		{
			return "0";
		}

		int n = s.length();
		int m = ( n + 1 ) / 2;
		char[] t = s.substring( 0, m ).toCharArray();
		
		for ( int i = m - 1; i >= 0; i-- )
		{
			if ( t[i] > '0' )
			{
				t[i]--;
				break;
			}
			else
			{
				t[i] = '9';
			}
		}

		if ( t[0] < '1' )
		{
			return "9".repeat( n - 1 );
		}

		String u = new String( t );

		return u + new StringBuilder( u.substring( 0, n - m ) ).reverse().toString();
	}

}
